### ISSUE TITLE
Members of frozen dataclass are considered invariant, but should not

### ISSUE THREAD

**Issue by igoras1993** [microsoft/pyright#8565]
## Bug description

Pyright considers `dataclass` members as invariant/mutable, even with `@dataclass(frozen=True)`. This **was fixed in `v1.1.329`** release, but also has been **reintroduced in `v1.1.339`**.

## Code

```python
from dataclasses import dataclass
from typing import Protocol


class A(Protocol): ...


class ComposeBase(Protocol):
    a: A


class A1(A): ...


@dataclass(frozen=True)
class Composed(ComposeBase):
    a: A1

```
- Interpreter: `Python 3.12.4`

### Result

```
poc.py
  poc.py:17:5 - error: "a" overrides symbol of same name in class "ComposeBase"
    Variable is mutable so its type is invariant
      Override type "A1" is not the same as base type "A" (reportIncompatibleVariableOverride)
1 error, 0 warnings, 0 informations 
```

### Command-line

- `pyright==1.1.329`: Bug **does not** exist
- `pyright==1.1.339`: Bug **does** exist
- `pyright==1.1.373` (recent): Bug **does** exists

### Pylance

In VSCode, `pylance` language server **does-not** report any in-IDE issues. 

#### VSCode / pylance versions

- ```shell
  $ code --version
  1.91.1
  f1e16e1e6214d7c44d078b1f0607b2388f29d729
  x64
  ```
- `Pylance` extension version: `v2024.7.1`

## Related issues

- #6029 
- #5993

**Comment by erictraut** [comment]
Pyright's behavior is correct here, so I don't consider this a bug.

The fact that the dataclass is frozen in this case is irrelevant. The error you're seeing is due to an override type violation. That `Composed` derives from the protocol `ComposedBase`, so it must honor the protocol definition. The protocol defines `a` as a writable attribute of type `A`, but you're attempting to override it with a read-only attribute of type `A1`. If you were to pass this object to a function that accepts a value of type `ComposedBase`, that function could write a value of type `A` to attribute `a`, which would violate the type of `Composed`.

There is currently no way to mark a protocol attribute as "read only". It has been proposed that the [ReadOnly type qualifier](https://typing.readthedocs.io/en/latest/spec/typeddict.html#typing-readonly-type-qualifier), introduced in PEP 705, could be used for this purpose. But no one has written a spec for this yet.

**Comment by igoras1993** [comment]
@erictraut Thank You for explanation!
